onnxruntime/test/python/onnxruntime_test_python_iobinding.py

```diff
@@ -42,31 +43,94 @@ def create_model_and_input(self):
         
-        return touch_input, torch_output
+        return touch_input, torch_output, torch_output_vals
 
     def test_bind_input_only(self):
-        x, torch_output = self.create_model_and_input()
+        torch_input, torch_output, torch_output_vals = self.create_model_and_input()
 
         session = onnxruntime.InferenceSession('model.onnx')
         io_binding = session.io_binding()
-        io_binding.bind_input('input', x.device.type, 0, np.float32, list(x.size()), x.data_ptr())
+        
+        # Bind input to CUDA
+        io_binding.bind_input('input', torch_input.device.type, 0, np.float32, list(torch_input.size()), torch_input.data_ptr())
+
+        # Bind output to CPU
         io_binding.bind_output('output')
+        
+        # Invoke Run
         session.run_with_iobinding(io_binding)
+        
+        # Get outputs over to CPU (the outputs which were bound to CUDA will get copied over to the host here)
         ort_output = io_binding.get_outputs()[0]
-    
-        self.assertTrue(np.array_equal(torch_output, ort_output))
+
+        # Validate results
+        self.assertTrue(np.array_equal(torch_output_vals, ort_output))
 
     def test_bind_input_to_cpu_arr(self):
-        torch_input, torch_output = self.create_model_and_input()
+        torch_input, torch_outpu, torch_output_vals = self.create_model_and_input()
 
         session = onnxruntime.InferenceSession('model.onnx')
         io_binding = session.io_binding()
+        
+        # Bind Numpy object (input) to CUDA
         io_binding.bind_cpu_input('input', torch_input.cpu().detach().numpy())
+        
+        # Bind output to CPU
         io_binding.bind_output('output')
+        
+        # Invoke Run
         session.run_with_iobinding(io_binding)
+        
+        # Get outputs over to CPU (the outputs which were bound to CUDA will get copied over to the host here)
         ort_output = io_binding.get_outputs()[0]
+
+        # Validate results
+        self.assertTrue(np.array_equal(torch_output_vals, ort_output))
+
+    def test_bind_input_and_preallocated_output(self):
+        torch_input, torch_output, torch_output_vals = self.create_model_and_input()
+
+        session = onnxruntime.InferenceSession('model.onnx')
+        io_binding = session.io_binding()
+
+        # Bind input to CUDA
+        io_binding.bind_input('input', torch_input.device.type, 0, np.float32, list(torch_input.size()), torch_input.data_ptr())
+
+        # Bind output to CUDA
+        ort_output = torch.empty(list(torch_output.size())).cuda()
+        io_binding.bind_output('output', torch_output.device.type, 0, np.float32, list(torch_output.size()), ort_output.data_ptr())
+
+        # Invoke Run
+        session.run_with_iobinding(io_binding)
+        
+        # Get outputs over to CPU (the outputs which were bound to CUDA will get copied over to the host here)
+        ort_output_vals = io_binding.get_outputs()[0]        
```

Comment:
#4415 had o`rt_output_vals = ort_output.cpu().detach().numpy()` to copy over the memory contents from CUDA to CPU and return a Numpy object. Felt it is meaningful to use `get_outputs()` and tests its functionality of making the device-host copy.

Reply:
ort_output.cpu().detach().numpy() must be tested too as that is the only way to make sure ORT actually wrote into the pre-allocated buffer. Thanks @skottmckay